Makefile

```diff
@@ -434,3 +405,96 @@ site-serve: .$(DOCKER_SITE_BUILDER_IMAGE).image.stamp
 	touch $@
 
 docs: site-build
+
+#
+# CRD et. al code generation
+#
+
+CRD_BASE_DIR := config/crd/bases
+HELM_TOP_DIR := deployment/helm
+
+# Location to install external generation tool binaries
+BUILD_AUX ?= $(shell pwd)/build-aux
+LOCALBIN ?= $(BUILD_AUX)/bin
+$(LOCALBIN):
+	mkdir -p $(LOCALBIN)
+
+## External generation tool binaries
+KUSTOMIZE ?= $(LOCALBIN)/kustomize
+CONTROLLER_GEN ?= $(LOCALBIN)/controller-gen
+CLIENT_GEN ?= $(LOCALBIN)/client-gen
+
+GIT_CLONE    := git clone
+GIT_CHECKOUT := git checkout
+
+# External generation tool versions
+CONTROLLER_TOOLS_VERSION ?= v0.12.0
+CLIENT_GEN_VERSION ?= v0.28.1
+CODE_GENERATOR_REPO ?= https://github.com/kubernetes/code-generator
+CODE_GENERATOR_VERSION ?= v0.28.1
+GENERATE_GROUPS ?= $(BUILD_AUX)/code-generator/generate-groups.sh
+
+# controller-gen local installation rules
+.PHONY: controller-gen
+controller-gen: $(CONTROLLER_GEN)
+$(CONTROLLER_GEN): $(LOCALBIN)
+	test -s $@ && $@ --version | grep -q $(CONTROLLER_TOOLS_VERSION) || \
+	GOBIN=$(LOCALBIN) go install sigs.k8s.io/controller-tools/cmd/controller-gen@$(CONTROLLER_TOOLS_VERSION)
+
+# client-gen local installation rules
+.PHONY: client-gen
+client-gen: $(CLIENT_GEN)
+$(CLIENT_GEN): $(LOCALBIN)
+	test -s $@ && $@ --version | grep -q $(CLIENT_GEN_VERSION) || \
+	GOBIN=$(LOCALBIN) go install k8s.io/code-generator/cmd/client-gen@$(CLIENT_GEN_VERSION)
+
+# code-generator local installation rules
+.PHONY: $(GENERATE_GROUPS)
+$(GENERATE_GROUPS):
+	test -s $@ || \
+	$(GIT_CLONE) $(CODE_GENERATOR_REPO) -b $(CODE_GENERATOR_VERSION) $(dir $@) && \
+	cd $(dir $@) && \
+	for d in ./cmd/*; do go build $$d; done
+
+# CRD generation rules
+.PHONY: generate-manifests
+generate-manifests: controller-gen
+	$(CONTROLLER_GEN) rbac:roleName=manager-role crd paths="./pkg/apis/..." output:crd:artifacts:config=$(CRD_BASE_DIR)
+
+# code generation rules for CRD types (DeepCopy, DeepCopyInto, etc.)
+.PHONY: generate-type
+generate-types: controller-gen
+	$(CONTROLLER_GEN) object:headerFile="./docs/license-header.go.txt" paths="./pkg/apis/..."
+
+# client generation rules 
+.PHONY: generate-clients
+generate-clients: $(GENERATE_GROUPS)
+	$(GENERATE_GROUPS) client \
+	    github.com/containers/nri-plugins/pkg/generated \
+	    github.com/containers/nri-plugins/pkg/apis \
```

Comment:
we don't have these directories here I guess? At least, `apis/` is under `pkg/`. Should we fix it here?

Reply:
Please ignore.